metrics/scaling/k8s_scale.sh

```diff
@@ -0,0 +1,267 @@
+#!/bin/bash
+# Copyright (c) 2019 Intel Corporation
+# 
+# SPDX-License-Identifier: Apache-2.0
+#
+
+set -e
+
+# Pull in some common, useful, items
+SCRIPT_PATH=$(dirname "$(readlink -f "$0")")
+source "${SCRIPT_PATH}/../lib/common.bash"
+
+input_yaml="bb.yaml.in"
+generated_yaml="generated.yaml"
+deployment="busybox"
+
+stats_pod="stats"
+
+export RUNTIME="kata-qemu"
+NUM_PODS=${NUM_PODS:-20}
+STEP=${STEP:-1}
+
+LABEL=${LABEL:-magiclabel}
+LABELVALUE=${LABELVALUE:-gandalf}
+
+# sleep and timeout times for k8s actions, in seconds
+wait_time=${wait_time:-30}
+delete_wait_time=${delete_wait_time:-600}
+settle_time=${settle_time:-5}
+use_kata_runtime=${use_kata_runtime:-no}
+
+# Set some default metrics env vars
+TEST_ARGS="runtime=${RUNTIME}"
+TEST_NAME="k8s scaling"
+
+get_num_nodes() {
+	n=$(kubectl get nodes --no-headers=true | wc -l)
+	echo "$n"
+}
+
+# $1 is the launch time in seconds this pod/container took to start up.
+# $2 is the number of pod/containers under test
+grab_stats() {
+	local launch_time=$1
+	local n_pods=$2
+	info "And grab some stats"
+	# Tell mpstat to measure over a short period, not only so we get slightly de-noised data, but also
+	# if you don't tell it the period, you will get the avg since boot, which is not what we want.
+	local cpu_idle=$(kubectl exec -ti ${stats_pod} -- sh -c "mpstat -u 3 1 | tail -1 | awk '{print \$11}'" | sed 's/\r//')
+	local mem_free=$(kubectl exec -ti ${stats_pod} -- sh -c "free | tail -2 | head -1 | awk '{print \$4}'" | sed 's/\r//')
+
+	info "idle [$cpu_idle] free [$mem_free] launch [$launch_time]"
+
+	# Annoyingly, it seems sometimes once in a while we don't get an answer!
+	# We should really retry, but for now, make the json valid at least
+	cpu_idle=${cpu_idle:-0}
+	mem_free=${mem_free:-0}
+
+	local json="$(cat << EOF
+	{
+		"n_pods": {
+			"Result": ${n_pods},
+			"Units" : "int"
+		},
+		"cpu_idle": {
+			"Result": ${cpu_idle},
+			"Units" : "%"
+		},
+		"launch_time": {
+			"Result": $launch_time,
+			"Units" : "s"
+		},
+		"mem_free": {
+			"Result": ${mem_free},
+			"Units" : "kb"
+		}
+	}
+EOF
+)"
+	metrics_json_add_array_element "$json"
+}
+
+init() {
+	info "Initialising"
+	info "Checking k8s accessible"
+	local worked=$( kubectl get nodes > /dev/null 2>&1 && echo $? || echo $? )
+	if [ "$worked" != 0 ]; then
+		die "kubectl failed to get nodes"
+	fi
+
+	info $(get_num_nodes) "k8s nodes found"
+	# We could check we have just the one node here - right now this is a single node
+	# test!! - because, our stats gathering is rudimentry, as k8s does not provide
+	# a nice way to do it (unless you want to parse 'descibe nodes')
+	# Have a read of https://github.com/kubernetes/kubernetes/issues/25353
+
+	# FIXME - check the node(s) can run enough pods - check 'max-pods' in the
+	# kubelet config - from 'kubectl describe node -o json' ?
+
+	# Launch our stats gathering pod
+	kubectl apply -f ${stats_pod}.yaml
```

Comment:
I've just found, we might want to prepend the `${SCRIPT_PATH}` to the front of the yaml files etc., as if you don't run the test from within its own directory right now it fails to find `stats.yaml` for instance.

Reply:
that will be true for bb.yaml.in too, it's mergeable without updating, but will fix

Reply:
updated for bb.yaml.in, generated.yaml, and stats.yaml